LibWeb: Fix "attempt to update a history entry's document"

dlarocque opened the pull request:
This updates the implementation of [attempt-to-populate-the-history-entry's-document](https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-populate-the-history-entry's-document) after fixes were made to the logic inside the queued task in https://github.com/whatwg/html/commit/cdd014ae84faf3165f8ce8441d132bb0e76513fb.

trflynn89 commented:
You're gonna want to bring your branch up to date:
![git](https://github.com/user-attachments/assets/f004d0f0-ce41-4b07-9a42-b6d423591af1)

dlarocque commented:
> You're gonna want to bring your branch up to date: ![git](https://private-user-images.githubusercontent.com/5600524/365446422-f004d0f0-ce41-4b07-9a42-b6d423591af1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU4MTUwNDQsIm5iZiI6MTcyNTgxNDc0NCwicGF0aCI6Ii81NjAwNTI0LzM2NTQ0NjQyMi1mMDA0ZDBmMC1jZTQxLTRiMDctOWE0Mi1iNmQ0MjM1OTFhZjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MDhUMTY1OTA0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmI2MmUwMTllNWEwMmQ1NzBjNTJlYTA1Y2Y1Njk4MmRkY2VmZDdiYWM1ZmExOTI4NmU5YWU1ZWNjZDgzNDZlNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.BLi0dGOBmbgk9rZjEZIzrxJmTZzWxznbOYTRgL_hEhI)

@trflynn89 Thanks for the reminder- should be good now

ADKaster commented on Userland/Libraries/LibWeb/HTML/Navigable.cpp:
```diff
@@ -1065,9 +1065,10 @@ WebIDL::ExceptionOr<void> Navigable::populate_session_history_entry_document(
+        // 2. Otherwise, if all of the following are true:
         //    - entry's URL's scheme is a fetch scheme; and
         //    - documentResource is null, or allowPOST is true and documentResource's request body is not failure (FIXME: check if request body is not failure)
+        // then set navigationParas to the result of creating navigation params by fetching given entry, navigable, sourceSnapshotParams, targetSnapshotParams, cspNavigationType, navigationId, and navTimingType.
```
is "navigationParas" your typo or Domenic's in the spec? 👀

dlarocque replied:
Oops, that's my typo- fixed.

ADKaster commented on Userland/Libraries/LibWeb/HTML/Navigable.cpp:
```diff
@@ -1152,57 +1140,48 @@ WebIDL::ExceptionOr<void> Navigable::populate_session_history_entry_document(
+            }
+            // FIXME: 3. If entry's document state's request referrer is "client", and navigationParams is a navigation params (i.e., neither null nor a non-fetch scheme navigation params), then:
+            //        https://github.com/whatwg/html/issues/9767
+            //        What is "request" here?
```
curious that the confusing reference to "Request" was not addressed. given the issue was closed 8 months ago we should probably open a new bug.

Unless the new steps do lay out who "request" is in this context?

dlarocque replied:
Looking at it now, would *"entry's document state's request referrer"* refer to `entry->document_state()->request_referrer()`?  I've updated my last commit to try to implement this step and remove the FIXME

dlarocque replied:
I added this conditional statement, but left the following steps as `FIXME`'s, since it looks like we can't yet assert that navigationParams's request is not null, since navigationParams may be set to a `AK::StringView` when there are network errors.

I'll look into filling in those assertions in a later PR.

ADKaster reviewed (commented):
thanks for cleaning this up in response to spec changes! two comments:

dlarocque commented:
Does anyone know whether the test failures in https://github.com/LadybirdBrowser/ladybird/actions/runs/10765631824/job/29850166244?pr=1330 are related to my changes? To me, they seem related to LibGfx.

trflynn89 commented:
I see this in one of the logs, so definitely related:
```
VERIFICATION FAILED: navigation_params.get<JS::NonnullGCPtr<NavigationParams>>()->request at /home/runner/work/ladybird/ladybird/Userland/Libraries/LibWeb/HTML/Navigable.cpp:1185
```

dlarocque commented:
> I see this in one of the logs, so definitely related:
> 
> ```
> VERIFICATION FAILED: navigation_params.get<JS::NonnullGCPtr<NavigationParams>>()->request at /home/runner/work/ladybird/ladybird/Userland/Libraries/LibWeb/HTML/Navigable.cpp:1185
> ```

Ah you're right- I removed those assertions from this PR and replaced them with `FIXME`'s (they were `FIXME`'s originally, so I'm not actually changing anything here).
